lib/Elevate/Components/DatabaseUpgrade.pm

```diff
@@ -26,15 +38,95 @@ sub pre_leapp ($self) {
     # If the database version is supported on the new OS version, then no need to upgrade
     return if Elevate::Database::is_database_version_supported( Elevate::Database::get_local_database_version() );
 
+    _ensure_localhost_mysql_profile_is_active();
+
     Elevate::Database::upgrade_database_server();
 
     return;
 }
 
 sub post_leapp ($self) {
 
+    if ( -e MYSQL_PROFILE_FILE ) {
+        my $original_profile = File::Slurp::read_text(MYSQL_PROFILE_FILE) // 'localhost';
+        INFO(qq{Re-activating "$original_profile" MySQL profile});
+        Cpanel::SafeRun::Simple::saferunnoerror( '/bin/sh', '-c', qq{/usr/local/cpanel/scripts/manage_mysql_profiles --activate $original_profile} );
+        unlink MYSQL_PROFILE_FILE;
+    }
+
     # Nothing to do
     return;
 }
 
+sub _ensure_localhost_mysql_profile_is_active {
+    my ($retry_count) = @_;
+
+    if ( Cpanel::MysqlUtils::MyCnf::Basic::is_local_mysql() ) {
+        return 1;
+    }
+
+    my $profile_manager = Cpanel::MysqlUtils::RemoteMySQL::ProfileManager->new();
+
+    # Validate that the current “localhost” profile exists, and contains valid settings.
+    try {
+        $profile_manager->validate_profile('localhost');
+        _activate_localhost_profile($profile_manager);
+    }
+
+    # Otherwise attempt to recreate it, overwriting the existing profile.
+    catch {
+        INFO("Attempting to create new localhost MySQL profile...");
+        if ( !$retry_count ) {
+            _create_new_localhost_profile($profile_manager);
+            _ensure_localhost_mysql_profile_is_active(1);
+        }
+        else {
+            die "Unable to generate/enable localhost MySQL profile: $_\n";
+        }
+    };
+
+    return;
+}
+
+sub _create_new_localhost_profile {
+    my ($profile_manager) = @_;
+
+    my $active_profile = $profile_manager->get_active_profile('dont_die');
+    File::Slurper::write_text( MYSQL_PROFILE_FILE, $active_profile );
+
+    my $password = Cpanel::PasswdStrength::Generate::generate_password( 16, no_othersymbols => 1 );
+
+    $profile_manager->create_profile(
+        {
+            'name'       => 'localhost',
+            'mysql_user' => 'root',
+            'mysql_pass' => $password,
+            'mysql_host' => 'localhost',
+            'mysql_port' => Cpanel::MysqlUtils::MyCnf::Basic::getmydbport('root') || 3306,
+            'setup_via'  => 'Auto-generated localhost profile during elevate.',
+        },
+        { 'overwrite' => 1 },
+    );
+
+    INFO("Resetting password for local root MySQL user...");
+    Cpanel::SafeRun::Simple::saferunnoerror( q{/bin/sh}, q{-c}, qq{/usr/local/cpanel/bin/whmapi1 set_local_mysql_root_password password='$password'} );
```

Comment:
FIX:  We should capture the output of the API call and make sure that it succeeds.  If the API call fails, we either need to retry or die here.